app/controllers/carts_controller.rb

```diff
@@ -4,8 +4,13 @@ class CartsController < ApplicationController
   before_action :require_shopper
 
   def show
-    flash[:primary] = "Your cart is empty." if @cart.total_count == 0
     set_cart
+    flash[:primary] = "Your cart is empty." if @cart.total_count == 0
+
+    if !current_user && @cart.total_count > 0
+
+      flash[:danger] = %Q[You must <a href="/login">login</a> or <a href="/register">register</a> to checkout.].html_safe
```

Comment:
%Q[Hello] is the same as "Hello". Just added that to then be able to add the double quotes in the link to make it clearer.

html_safe is because, without it, the view will literally print that exact string, it escapes the html.

Reply:
Nice work!